src/python/pants/core/util_rules/asdf.py

```diff
@@ -0,0 +1,238 @@
+        # Standardize that the last line for each tool wins.
+        #
+        # The definition of a tool-versions file can be found here:
+        # https://asdf-vm.com/#/core-configuration?id=tool-versions
+        tool_versions_lines = tool_versions_file.read_text().splitlines()
+        last_line = None
+        for line in tool_versions_lines:
+            # Find the last line for this tool.
+            if line.lower().startswith(tool_name):
+                last_line = line
+        if last_line:
+            _, _, versions = last_line.partition(tool_name)
+            for v in re.split(r"\s+", versions.strip()):
+                if ":" in v:
+                    key, _, value = v.partition(":")
+                    if key.lower() == "path":
+                        asdf_paths.append(value)
+                    elif key.lower() == "ref":
+                        asdf_versions[value] = str(tool_versions_file)
+                    else:
+                        logger.warning(
+                            f"Unknown version format `{v}` from ASDF configured by "
+                            f"`{paths_option_name}`, ignoring. This "
+                            f"version will not be considered when determining which {tool_description} "
+                            f"to use. Please check that `{tool_versions_file}` "
+                            "is accurate."
+                        )
+                elif v == "system":
+                    logger.warning(
+                        f"System path set by ASDF configured by `{paths_option_name}` is unsupported, ignoring. "
+                        f"This version will not be considered when determining which {tool_description} to use. "
+                        f"Please remove 'system' from `{tool_versions_file}` to disable this warning."
+                    )
+                else:
+                    asdf_versions[v] = str(tool_versions_file)
+
+    for version, source in asdf_versions.items():
+        install_dir = asdf_installs_dir / version / "bin"
+        if install_dir.exists():
+            asdf_paths.append(str(install_dir))
+        else:
+            logger.warning(
+                f"Trying to use ASDF version `{version}` configured by "
+                f"`{paths_option_name}` but `{install_dir}` does not "
+                f"exist. This version will not be considered when determining which {tool_description} "
+                f"to use. Please check that `{source}` is accurate."
+            )
+
+    # For non-local, if no paths have been defined, fallback to every version installed
+    if not local and len(asdf_paths) == 0:
+        # This could be appended to asdf_paths, but there isn't any reason to
+        return tuple(asdf_installed_paths)
+    else:
+        return tuple(asdf_paths)
+
+
+@rule
+async def resolve_asdf_tool_paths(
+    request: AsdfToolPathsRequest, build_root: BuildRoot
+) -> AsdfToolPathsResult:
```

Comment:
Given the direct HOME and path access in the `@rule_helper`, can you mark this `@_uncacheable_rule`? I think that it's been a longstanding issue that ASDF edits won't invalidate correctly.

Reply:
Is there those path checks could use `PathGlobs` or an equivalent? I tried using PathGlobs in an earlier iteration of this PR to read `.tool-versions`, but the default `--pants-ignore` option ignores dot files in the build root.

Reply:
(Any fix to be able to use `PathGlobs` would be for the future, not this PR.)